feat: gitmodules for contracts deps

tynes opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

Migrates away from using npm for contracts deps to git submodules. As much as people like to complain about git submodules, it is the best that we have and the industry standard for foundry projects. The main problem this is solving is not needing to have the Go tests depend on the installation of the JS deps to work

tynes committed:
forge install: openzeppelin-contracts-upgradeable

v4.7.3

tynes committed:
forge install: openzeppelin-contracts

v4.7.3

tynes committed:
forge install: solmate

8f9b23f8838670afda0fd8983f2c41e8037ae6bc

tynes committed:
forge install: clones-with-immutable-args

105efee1b9127ed7f6fedf139e1fc796ce8791f2

tynes committed:
forge install: ds-test

c9ce3f25bde29fc5eb9901842bf02850dfd2d084

tynes committed:
forge install: forge-std

e8a047e3f40f13fa37af6fe14e6e06283d9a060e

tynes committed:
feat: git submodules

refcell commented on packages/contracts-bedrock/package.json:
```diff
@@ -39,17 +39,10 @@
-    "@rari-capital/solmate": "github:transmissions11/solmate#8f9b23f8838670afda0fd8983f2c41e8037ae6bc",
-    "clones-with-immutable-args": "github:Saw-mon-and-Natalie/clones-with-immutable-args#105efee1b9127ed7f6fedf139e1fc796ce8791f2"
-  },
+  "dependencies": {},
```
Can we just remove the dependencies kv pair?

tynes committed:
dockerfile: add .gitmodules for foundry

tynes committed:
contracts-bedrock: clean up package.json

tynes committed:
ci: cleanup

dockerfile: fix typo

docker: attempt fix

ci: fix slither

ci: fix deps

docker: fix

op-bindings: regenerate

ci: build first

ci: fixup

refcell reviewed (approved):
I believe this looks good to me now

tynes commented:
Need to figure out the issue with the docker build

tynes committed:
op-bindings: regenerate

tynes commented:
This unblocks https://github.com/ethereum-optimism/optimism/pull/6412 and speeds up contracts CI as it no longer depends on the typescript install for running tests/coverage/slither. The typescript deps still are present for the checks, that is something that we could remove in the future so that the entire contracts CI has no steps that it depends on

tynes commented on .circleci/config.yml:
```diff
@@ -502,34 +456,12 @@ jobs:
           command: |
-            slither --version && pnpm slither || echo "export LINT_STATUS=1" >> "$BASH_ENV"
-      - run:
-          name: invariant docs
```
This is done twice, no need to do it twice, its also in `contracts-bedrock-checks`

refcell replied:
See: https://github.com/ethereum-optimism/optimism/blob/b0e3ed19441df4e0d272fbda1b7f9df8b056d9b1/.circleci/config.yml#L420-L425

tynes commented on .circleci/config.yml:
```diff
@@ -502,34 +456,12 @@ jobs:
       - run:
           name: slither
           command: |
-            slither --version && pnpm slither || echo "export LINT_STATUS=1" >> "$BASH_ENV"
```
The or case here can fail silently, preserving that behavior but no need to write something to the env in this case

refcell reviewed (approved):
Re-reviewed, looks good to me!

Notice for others, the invariant-docs ci check removal is a de-dup (see 2nd instance [here](https://github.com/ethereum-optimism/optimism/blob/b0e3ed19441df4e0d272fbda1b7f9df8b056d9b1/.circleci/config.yml#L420-L425))

maurelian reviewed (approved):
LGTM

Can we also just bite the bullet and move `contracts -> src` now?

tynes commented:
> LGTM
> 
> Can we also just bite the bullet and move `contracts -> src` now?

Can follow up with another PR for this soon
